File: test/python-tests/test_advanced_decorators.py
#!/usr/bin/env python3
"""
Advanced Python Decorator Patterns
Including class decorators, parameterized decorators, and decorator factories
"""

import functools
import time
import logging
import threading
from typing import Any, Callable, Type, TypeVar, cast

F = TypeVar('F', bound=Callable[..., Any])
T = TypeVar('T')

# Class Decorators

def singleton(cls: Type[T]) -> Type[T]:
    """Class decorator to implement singleton pattern"""
    instances = {}
    
    @functools.wraps(cls)
    def wrapper(*args, **kwargs):
        if cls not in instances:
            instances[cls] = cls(*args, **kwargs)
        return instances[cls]
    
    return cast(Type[T], wrapper)

def count_instances(cls: Type[T]) -> Type[T]:
    """Class decorator to count instances"""
    cls._instance_count = 0
    original_init = cls.__init__
    
    @functools.wraps(original_init)
    def new_init(self, *args, **kwargs):
        cls._instance_count += 1
        original_init(self, *args, **kwargs)
    
    cls.__init__ = new_init
    
    @classmethod
    def get_instance_count(cls):
        return cls._instance_count
    
    cls.get_instance_count = get_instance_count
    return cls

def auto_str(cls: Type[T]) -> Type[T]:
    """Class decorator to automatically implement __str__ method"""
    def __str__(self):
        attributes = []
        for attr, value in self.__dict__.items():
            attributes.append(f"{attr}={value!r}")
        return f"{cls.__name__}({', '.join(attributes)})"
    
    cls.__str__ = __str__
    return cls

# Parameterized Decorators

def retry(max_attempts: int = 3, delay: float = 1.0, backoff: float = 2.0):
    """Parameterized decorator for retry logic with exponential backoff"""
    def decorator(func: F) -> F:
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            current_delay = delay
            last_exception = None
            
            for attempt in range(max_attempts):
                try:
                    return func(*args, **kwargs)
                except Exception as e:
                    last_exception = e
                    if attempt == max_attempts - 1:
                        raise
                    
                    logging.warning(f"Attempt {attempt + 1} failed: {e}. Retrying in {current_delay}s...")
                    time.sleep(current_delay)
                    current_delay *= backoff
            
            raise last_exception
        
        return cast(F, wrapper)
    return decorator

def timeout(seconds: float):
    """Parameterized decorator to add timeout to function execution"""
    def decorator(func: F) -> F:
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            result = [None]
            exception = [None]
            
            def target():
                try:
                    result[0] = func(*args, **kwargs)
                except Exception as e:
                    exception[0] = e
            
            thread = threading.Thread(target=target)
            thread.daemon = True
            thread.start()
            thread.join(timeout=seconds)
            
            if thread.is_alive():
                raise TimeoutError(f"Function {func.__name__} timed out after {seconds} seconds")
            
            if exception[0]:
                raise exception[0]
            
            return result[0]
        
        return cast(F, wrapper)
    return decorator

def cache_with_ttl(ttl_seconds: float = 300):
    """Parameterized decorator for caching with TTL (time to live)"""
    def decorator(func: F) -> F:
        cache = {}
        cache_timestamps = {}
        
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            # Create cache key from arguments
            cache_key = str(args) + str(sorted(kwargs.items()))
            current_time = time.time()
            
            # Check if cached result exists and is not expired
            if cache_key in cache:
                if current_time - cache_timestamps[cache_key] < ttl_seconds:
                    print(f"Cache hit for {func.__name__}{args}")
                    return cache[cache_key]
                else:
                    # Remove expired entry
                    del cache[cache_key]
                    del cache_timestamps[cache_key]
            
            # Compute and cache result
            print(f"Cache miss for {func.__name__}{args}")
            result = func(*args, **kwargs)
            cache[cache_key] = result
            cache_timestamps[cache_key] = current_time
            
            return result
        
        # Add cache management methods
        def clear_cache():
            cache.clear()
            cache_timestamps.clear()
            print(f"Cache cleared for {func.__name__}")
        
        def cache_info():
            return {
                'size': len(cache),
                'keys': list(cache.keys())
            }
        
        wrapper.clear_cache = clear_cache
        wrapper.cache_info = cache_info
        
        return cast(F, wrapper)
    return decorator

def rate_limit(max_calls: int = 10, time_window: float = 60.0):
    """Parameterized decorator for rate limiting function calls"""
    def decorator(func: F) -> F:
        calls = []
        lock = threading.Lock()
        
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            current_time = time.time()
            
            with lock:
                # Remove calls outside the time window
                calls[:] = [call_time for call_time in calls if current_time - call_time < time_window]
                
                # Check if we've exceeded the rate limit
                if len(calls) >= max_calls:
                    oldest_call = min(calls)
                    wait_time = time_window - (current_time - oldest_call)
                    raise RuntimeError(f"Rate limit exceeded. Wait {wait_time:.1f} seconds.")
                
                # Record this call
                calls.append(current_time)
            
            return func(*args, **kwargs)
        
        return cast(F, wrapper)
    return decorator

# Advanced Decorator Patterns

def conditional(condition: Callable[[], bool]):
    """Decorator that only applies the decorated function if condition is true"""
    def decorator(func: F) -> F:
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            if condition():
                return func(*args, **kwargs)
            else:
                print(f"Skipping {func.__name__} - condition not met")
                return None
        
        return cast(F, wrapper)
    return decorator

def compose(*decorators):
    """Compose multiple decorators into one"""
    def decorator(func: F) -> F:
        for dec in reversed(decorators):
            func = dec(func)
        return func
    return decorator

def logging_decorator(logger_name: str = None, level: int = logging.INFO):
    """Advanced logging decorator with configurable logger and level"""
    def decorator(func: F) -> F:
        logger = logging.getLogger(logger_name or func.__module__)
        
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            start_time = time.time()
            logger.log(level, f"Calling {func.__name__} with args: {args}, kwargs: {kwargs}")
            
            try:
                result = func(*args, **kwargs)
                execution_time = time.time() - start_time
                logger.log(level, f"{func.__name__} completed successfully in {execution_time:.4f}s")
                return result
            except Exception as e:
                execution_time = time.time() - start_time
                logger.log(logging.ERROR, f"{func.__name__} failed after {execution_time:.4f}s: {e}")
                raise
        
        return cast(F, wrapper)
    return decorator

# Test Classes and Functions

@singleton
class DatabaseConnection:
    """Singleton database connection class"""
    def __init__(self, connection_string: str):
        self.connection_string = connection_string
        self.connected = True
        print(f"Database connection established: {connection_string}")
    
    def query(self, sql: str):
        return f"Executing query: {sql}"

@count_instances
@auto_str
class Person:
    """Person class with automatic string representation and instance counting"""
    def __init__(self, name: str, age: int):
        self.name = name
        self.age = age

class Calculator:
    """Calculator class for testing decorators"""
    
    @retry(max_attempts=3, delay=0.1)
    def divide_with_retry(self, a: float, b: float) -> float:
        """Division with automatic retry on failure"""
        print(f"Attempting division: {a} / {b}")
        if b == 0:
            raise ValueError("Division by zero")
        return a / b
    
    @timeout(2.0)
    def slow_operation(self, duration: float) -> str:
        """Simulate a slow operation"""
        print(f"Starting slow operation for {duration} seconds...")
        time.sleep(duration)
        return f"Operation completed after {duration} seconds"
    
    @cache_with_ttl(ttl_seconds=1.0)
    def expensive_calculation(self, n: int) -> int:
        """Expensive calculation with caching"""
        print(f"Performing expensive calculation for n={n}")
        time.sleep(0.5)  # Simulate expensive operation
        return n * n + n
    
    @rate_limit(max_calls=3, time_window=5.0)
    def api_call(self, endpoint: str) -> str:
        """Simulated API call with rate limiting"""
        print(f"Making API call to: {endpoint}")
        return f"Response from {endpoint}"

# Conditional execution flag
enable_debug_mode = False

def debug_mode_enabled():
    """Condition function for conditional decorator"""
    return enable_debug_mode

@conditional(debug_mode_enabled)
def debug_function():
    """Function that only runs in debug mode"""
    print("Debug mode is enabled - executing debug function")

# Composed decorators
@compose(
    logging_decorator(),
    retry(max_attempts=3, delay=0.1)
)
def complex_operation(fail_count: int) -> str:
    """Complex operation with logging and retry"""
    global _attempt_count
    if '_attempt_count' not in globals():
        _attempt_count = 0
    
    _attempt_count += 1
    if _attempt_count <= fail_count:
        raise RuntimeError(f"Simulated failure #{_attempt_count}")
    return f"Operation succeeded on attempt #{_attempt_count}"

def test_class_decorators():
    """Test class decorators"""
    print("=== Testing Class Decorators ===")
    
    # Test singleton decorator
    print("\n--- Singleton Decorator ---")
    db1 = DatabaseConnection("postgresql://localhost:5432/mydb")
    db2 = DatabaseConnection("postgresql://localhost:5432/otherdb")  # Should return same instance
    print(f"db1 is db2: {db1 is db2}")  # Should be True
    print(f"Connection string: {db1.connection_string}")
    
    # Test instance counting decorator
    print("\n--- Instance Counting Decorator ---")
    person1 = Person("Alice", 25)
    person2 = Person("Bob", 30)
    person3 = Person("Charlie", 35)
    
    print(f"Person instances created: {Person.get_instance_count()}")
    print(f"Person 1: {person1}")
    print(f"Person 2: {person2}")
    print(f"Person 3: {person3}")

def test_parameterized_decorators():
    """Test parameterized decorators"""
    print("\n=== Testing Parameterized Decorators ===")
    
    calc = Calculator()
    
    # Test retry decorator
    print("\n--- Retry Decorator ---")
    try:
        result = calc.divide_with_retry(10, 2)
        print(f"Division result: {result}")
        
        result = calc.divide_with_retry(10, 0)  # This will retry and eventually fail
    except ValueError as e:
        print(f"Division failed after retries: {e}")
    
    # Test timeout decorator
    print("\n--- Timeout Decorator ---")
    try:
        result = calc.slow_operation(1.0)  # Should succeed
        print(f"Slow operation result: {result}")
        
        result = calc.slow_operation(3.0)  # Should timeout
    except TimeoutError as e:
        print(f"Operation timed out: {e}")
    
    # Test cache with TTL decorator
    print("\n--- Cache with TTL Decorator ---")
    result1 = calc.expensive_calculation(5)
    print(f"First call result: {result1}")
    
    result2 = calc.expensive_calculation(5)  # Should use cache
    print(f"Second call result (cached): {result2}")
    
    print(f"Cache info: {calc.expensive_calculation.cache_info()}")
    
    # Wait for cache to expire
    print("Waiting for cache to expire...")
    time.sleep(1.5)
    
    result3 = calc.expensive_calculation(5)  # Should recalculate
    print(f"Third call result (after cache expiry): {result3}")
    
    # Test rate limiting decorator
    print("\n--- Rate Limiting Decorator ---")
    try:
        for i in range(4):  # Should succeed for first 3, fail on 4th
            result = calc.api_call(f"/api/endpoint{i}")
            print(f"API call {i+1}: {result}")
    except RuntimeError as e:
        print(f"Rate limit exceeded: {e}")

def test_advanced_patterns():
    """Test advanced decorator patterns"""
    print("\n=== Testing Advanced Decorator Patterns ===")
    
    # Test conditional decorator
    print("\n--- Conditional Decorator ---")
    global enable_debug_mode
    
    enable_debug_mode = False
    debug_function()  # Should be skipped
    
    enable_debug_mode = True
    debug_function()  # Should execute
    
    # Test composed decorators
    print("\n--- Composed Decorators ---")
    global _attempt_count
    _attempt_count = 0
    
    result = complex_operation(2)  # Should fail twice, then succeed on 3rd attempt
    print(f"Complex operation result: {result}")

def main():
    """Main test function"""
    print("=== Advanced Python Decorator Patterns ===")
    
    test_class_decorators()
    test_parameterized_decorators()
    test_advanced_patterns()
    
    print("\n=== All decorator tests completed successfully! ===")

if __name__ == "__main__":
    # Set up basic logging
    logging.basicConfig(level=logging.INFO, format='%(levelname)s: %(message)s')
    main()

File: test/python-tests/test_advanced_decorators.cpp
#include <iostream>
#include <string>
#include <vector>
#include <optional>
#include <unordered_map>
#include <tuple>
#include <utility>
#include <type_traits>
#include <functional>
#include <sstream>
#include <iomanip>
#include <cmath>
#include <algorithm>
#include <typeinfo>
#include <cstring>

using namespace std;

// helpers
template<typename T> static inline std::ostream& operator<<(std::ostream& os, const std::vector<T>& v);
template<typename T> static inline std::string to_str(const T& v){ std::ostringstream os; os<<v; return os.str(); }
static inline std::string to_str(const std::string& s){ return s; }
static inline std::string to_str(const char* s){ return std::string(s); }
static inline std::string to_str(double v){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(1)<<v; return os.str(); }
static inline std::string to_str(float v){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(1)<<v; return os.str(); }
template<typename T> static inline int to_int(const T& v){ if constexpr (std::is_convertible_v<T,std::string>) return std::stoi(v); else return static_cast<int>(v); }
template<typename T> static inline float to_float(const T& v){ if constexpr (std::is_convertible_v<T,std::string>) return std::stof(v); else return static_cast<float>(v); }
template<typename C> static inline int sum(const C& c){ int s=0; for(const auto& e: c) s += e; return s; }
template<typename T> static inline int sum(std::initializer_list<T> c){ int s=0; for(const auto& e: c) s += e; return s; }
template<typename T> static inline size_t len(const T& c){ return c.size(); }
static inline size_t len(const char* s){ return std::char_traits<char>::length(s); }
template<typename K, typename V> static inline std::string to_str(const std::unordered_map<K,V>&){ return std::string("{...}"); }
// minimal ABC base to support 'class X : public ABC'
struct ABC {};
// minimal math namespace replacement for 'math.pi'
struct __fluxus_math { static constexpr double pi = 3.141592653589793; static inline double sqrt(double x){ return std::sqrt(x); } } math;
// minimal sys module stub
namespace sys { static std::vector<std::string> argv; }
namespace asyncio { template<typename F> static inline void run(F f){ f(); } }

// vector to string conversion for printing
template<typename T>
static inline std::string vec_to_str(const std::vector<T>& v) {
    std::ostringstream os;
    os << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i > 0) os << ", ";
        if constexpr (std::is_same_v<T, std::string>) {
            os << "'" << v[i] << "'";
        } else {
            os << v[i];
        }
    }
    os << "]";
    return os.str();
}

// nested vector to string
template<typename T>
static inline std::string vec_to_str(const std::vector<std::vector<T>>& v) {
    std::ostringstream os;
    os << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i > 0) os << ", ";
        os << vec_to_str(v[i]);
    }
    os << "]";
    return os.str();
}

// stream operator for vectors
template<typename T>
static inline std::ostream& operator<<(std::ostream& os, const std::vector<T>& v) {
    os << vec_to_str(v);
    return os;
}
template<typename K, typename V> static inline std::ostream& operator<<(std::ostream& os, const std::unordered_map<K,V>&){ os << "{...}"; return os; }
template<typename K, typename V> static inline auto py_items(const std::unordered_map<K,V>& m){ std::vector<std::pair<K,V>> v; v.reserve(m.size()); for(const auto& kv: m){ v.emplace_back(kv.first, kv.second);} return v; }


class DatabaseConnection {
    public:
        DatabaseConnection(std::string connection_string: str) {
            this->connection_string = connection_string;
            this->connected = true;
        }
        std::string query(str sql) {
    return to_str("Executing query: ") + to_str(sql);
}


    public:
        int connection_string = 0;
        int connected = 0;
};


class Person {
    public:
        Person(std::string name: str, int age: int) {
            this->name = name;
            this->age = age;
        }
        void placeholder();
    public:
        std::string name = "";
        int age = 0;
};


class Calculator {
    public:
        Calculator() = default;
        int divide_with_retry(float_ a, float_ b) {
    /* docstring */
    std::cout << to_str("Attempting division: ") << to_str(a) << to_str(" / ") << to_str(b) << std::endl;
    if (b == 0) {
        /* raise */
    }
    return a / b;
}


        std::string slow_operation(float_ duration) {
    /* docstring */
    std::cout << to_str("Starting slow operation for ") << to_str(duration) << to_str(" seconds...") << std::endl;
    time.sleep(duration);
    return to_str("Operation completed after ") + to_str(duration) + to_str(" seconds");
}


        int expensive_calculation(int_ n) {
    /* docstring */
    std::cout << to_str("Performing expensive calculation for n=") << to_str(n) << std::endl;
    time.sleep(0.5);
    return n * n + n;
}


        std::string api_call(str endpoint) {
    /* docstring */
    std::cout << to_str("Making API call to: ") << to_str(endpoint) << std::endl;
    return to_str("Response from ") + to_str(endpoint);
}


    public:
        // generated placeholder fields
};


int singleton(Type[T] cls) {
    /* docstring */
    auto instances = {};
    @functools.wraps(cls);
    std::function<int()> wrapper = [&]() {
        if (cls not in instances) {
            auto instances[cls] = cls(*args, **kwargs);
        }
        return instances[cls];
    };
    return cast(Type[T], wrapper);
}


int count_instances(Type[T] cls) {
    /* docstring */
    cls._instance_count(0);
    auto original_init = cls.__init__;
    @functools.wraps(original_init);
    std::function<int()> new_init = [&]() {
        cls._instance_count += 1;
        original_init(self, *args, **kwargs);
    };
    cls.__init__(new_init);
    @classmethod;
    std::function<int()> get_instance_count = [&]() {
        return cls._instance_count;
    };
    cls.get_instance_count(get_instance_count);
    return cls;
}


std::string auto_str(Type[T] cls) {
    /* docstring */
    std::function<int()> __str__ = [&]() {
        auto attributes = std::vector<int>{};
        for (auto [attr, value] : self.py_items(__dict__)) {
            attributes.push_back(f"{attr}={value!r}");
        }
        return to_str(cls) + to_str("(") + to_str(', '.join(attributes)) + to_str(")");
    };
    cls.__str__(__str__);
    return cls;
}


std::function<int()> retry(int = 3 max_attempts, float = 1.0 delay, float = 2.0 backoff) {
    bool __fluxus_exc=false;
    /* docstring */
    std::function<int()> decorator = [&]() {
        @functools.wraps(func);
        std::function<int()> wrapper = [&]() {
            auto current_delay = delay;
            auto last_exception = 0;
            for (int attempt = 0; attempt < max_attempts; ++attempt) {
                {
                    return func(*args, **kwargs);
                }
                if (__fluxus_exc) { auto e = 0;
                    auto last_exception = e;
                    if (attempt == max_attempts - 1) {
                        raise;
                    }
                    logging.warning(f"Attempt {attempt + 1} failed: {e}. Retrying in {current_delay}s...");
                    time.sleep(current_delay);
                    current_delay *= backoff;
                }
            }
            /* raise */
        };
        return cast(F, wrapper);
    };
    return decorator;
}


std::function<int()> timeout(float_ seconds) {
    bool __fluxus_exc=false;
    /* docstring */
    std::function<int()> decorator = [&]() {
        @functools.wraps(func);
        std::function<int()> wrapper = [&]() {
            auto result = std::vector<decltype(0)>{0};
            auto exception = std::vector<decltype(0)>{0};
            std::function<int()> target = [&]() {
                {
                    auto result[0] = func(*args, **kwargs);
                }
                if (__fluxus_exc) { auto e = 0;
                    auto exception[0] = e;
                }
            };
            auto thread = threading.Thread(target=target);
            thread.daemon(true);
            thread.start();
            thread.join(timeout(seconds));
            if (thread.is_alive()) {
                /* raise */
            }
            if (exception[0]) {
                /* raise */
            }
            return result[0];
        };
        return cast(F, wrapper);
    };
    return decorator;
}


std::function<int()> cache_with_ttl(float = 300 ttl_seconds) {
    /* docstring */
    std::function<int()> decorator = [&]() {
        auto cache = {};
        auto cache_timestamps = {};
        @functools.wraps(func);
        std::function<int()> wrapper = [&]() {
            auto cache_key = to_str(args) + to_str(sorted(py_items(kwargs)));
            auto current_time = time.time();
            if (cache_key in cache) {
                if (current_time - cache_timestamps[cache_key] < ttl_seconds) {
                    std::cout << to_str("Cache hit for ") << to_str(func) << to_str(args) << std::endl;
                    return cache[cache_key];
                }
                else {
                    del cache[cache_key];
                    del cache_timestamps[cache_key];
                }
            }
            std::cout << to_str("Cache miss for ") << to_str(func) << to_str(args) << std::endl;
            auto result = func(*args, **kwargs);
            auto cache[cache_key] = result;
            auto cache_timestamps[cache_key] = current_time;
            return result;
        };
        std::function<int()> clear_cache = [&]() {
            cache.clear();
            cache_timestamps.clear();
            std::cout << to_str("Cache cleared for ") << to_str(func) << std::endl;
        };
        std::function<int()> cache_info = [&]() {
            return {;
                'size': len(cache),;
                'keys': list(cache.keys());
            };
            };
        };
        wrapper.clear_cache(clear_cache);
        wrapper.cache_info(cache_info);
        return cast(F, wrapper);
    }
    return decorator;
}


std::function<int()> rate_limit(int = 10 max_calls, float = 60.0 time_window) {
    /* docstring */
    std::function<int()> decorator = [&]() {
        auto calls = std::vector<int>{};
        auto lock = threading.Lock();
        @functools.wraps(func);
        std::function<int()> wrapper = [&]() {
            auto current_time = time.time();
            {
                auto calls[:] = 0;
                if (len(calls) >= max_calls) {
                    auto oldest_call = min(calls);
                    auto wait_time = time_window - (current_time - oldest_call);
                    /* raise */
                }
                calls.push_back(current_time);
            }
            return func(*args, **kwargs);
        };
        return cast(F, wrapper);
    };
    return decorator;
}


std::function<int()> conditional(Callable[[] condition, int bool]) {
    /* docstring */
    std::function<int()> decorator = [&]() {
        @functools.wraps(func);
        std::function<int()> wrapper = [&]() {
            if (condition()) {
                return func(*args, **kwargs);
            }
            else {
                std::cout << to_str("Skipping ") << to_str(func) << to_str(" - condition not met") << std::endl;
                return 0;
            }
        };
        return cast(F, wrapper);
    };
    return decorator;
}


std::function<int()> compose(std::initializer_list<int> decorators) {
    /* docstring */
    std::function<int()> decorator = [&]() {
        for (auto dec : reversed(decorators)) {
            auto func = dec(func);
        }
        return func;
    };
    return decorator;
}

template<typename... Ts>
auto compose(Ts... xs){ return compose(std::initializer_list<int>{ static_cast<int>(xs)... }); }


std::function<int()> logging_decorator(str = None logger_name, int = logging.INFO level) {
    bool __fluxus_exc=false;
    /* docstring */
    std::function<int()> decorator = [&]() {
        auto logger = logging.getLogger(logger_name or func.__module__);
        @functools.wraps(func);
        std::function<int()> wrapper = [&]() {
            auto start_time = time.time();
            logger.log(level, f"Calling {func} with args: {args}, kwargs: {kwargs}");
            {
                auto result = func(*args, **kwargs);
                auto execution_time = time.time() - start_time;
                logger.log(level, f"{func} completed successfully in {execution_time:.4f}s");
                return result;
            }
            if (__fluxus_exc) { auto e = 0;
                auto execution_time = time.time() - start_time;
                logger.log(logging.ERROR, f"{func} failed after {execution_time:.4f}s: {e}");
                raise;
            }
        };
        return cast(F, wrapper);
    };
    return decorator;
}


int debug_mode_enabled() {
    /* docstring */
    return enable_debug_mode;
}


void debug_function() {
    /* docstring */
    std::cout << "Debug mode is enabled - executing debug function" << std::endl;
}


std::string complex_operation(int_ fail_count) {
    /* docstring */
    global _attempt_count;
    if ('_attempt_count' not in globals()) {
        auto _attempt_count = 0;
    }
    _attempt_count += 1;
    if (_attempt_count <= fail_count) {
        /* raise */
    }
    return to_str("Operation succeeded on attempt #") + to_str(_attempt_count);
}


void test_class_decorators() {
    /* docstring */
    std::cout << "=== Testing Class Decorators ===" << std::endl;
    std::cout << "\n--- Singleton Decorator ---" << std::endl;
    auto db1 = DatabaseConnection("postgresql://localhost:5432/mydb");
    auto db2 = DatabaseConnection("postgresql://localhost:5432/otherdb");
    std::cout << to_str("db1 is db2: ") << to_str(db1 is db2) << std::endl;
    std::cout << to_str("Connection string: ") << to_str(db1.connection_string) << std::endl;
    std::cout << "\n--- Instance Counting Decorator ---" << std::endl;
    auto person1 = Person("Alice", 25);
    auto person2 = Person("Bob", 30);
    auto person3 = Person("Charlie", 35);
    std::cout << to_str("Person instances created: ") << to_str(Person.get_instance_count()) << std::endl;
    std::cout << to_str("Person 1: ") << to_str(person1) << std::endl;
    std::cout << to_str("Person 2: ") << to_str(person2) << std::endl;
    std::cout << to_str("Person 3: ") << to_str(person3) << std::endl;
}


void test_parameterized_decorators() {
    bool __fluxus_exc=false;
    /* docstring */
    std::cout << "\n=== Testing Parameterized Decorators ===" << std::endl;
    auto calc = Calculator();
    std::cout << "\n--- Retry Decorator ---" << std::endl;
    {
        auto result = calc.divide_with_retry(10, 2);
        std::cout << to_str("Division result: ") << to_str(result) << std::endl;
        auto result = calc.divide_with_retry(10, 0);
    }
    if (__fluxus_exc) { auto e = 0;
        std::cout << to_str("Division failed after retries: ") << to_str(e) << std::endl;
    }
    std::cout << "\n--- Timeout Decorator ---" << std::endl;
    {
        auto result = calc.slow_operation(1.0);
        std::cout << to_str("Slow operation result: ") << to_str(result) << std::endl;
        auto result = calc.slow_operation(3.0);
    }
    if (__fluxus_exc) { auto e = 0;
        std::cout << to_str("Operation timed out: ") << to_str(e) << std::endl;
    }
    std::cout << "\n--- Cache with TTL Decorator ---" << std::endl;
    auto result1 = calc.expensive_calculation(5);
    std::cout << to_str("First call result: ") << to_str(result1) << std::endl;
    auto result2 = calc.expensive_calculation(5);
    std::cout << to_str("Second call result (cached): ") << to_str(result2) << std::endl;
    std::cout << to_str("Cache info: ") << to_str(calc.expensive_calculation.cache_info()) << std::endl;
    std::cout << 0 << std::endl;
    time.sleep(1.5);
    auto result3 = calc.expensive_calculation(5);
    std::cout << to_str("Third call result (after cache expiry): ") << to_str(result3) << std::endl;
    std::cout << "\n--- Rate Limiting Decorator ---" << std::endl;
    {
        for (int i = 0; i < 4; ++i) {
            auto result = calc.api_call(f"/api/endpoint{i}");
            std::cout << to_str("API call ") << to_str(i+1) << to_str(": ") << to_str(result) << std::endl;
        }
    }
    if (__fluxus_exc) { auto e = 0;
        std::cout << to_str("Rate limit exceeded: ") << to_str(e) << std::endl;
    }
}


void test_advanced_patterns() {
    /* docstring */
    std::cout << "\n=== Testing Advanced Decorator Patterns ===" << std::endl;
    std::cout << "\n--- Conditional Decorator ---" << std::endl;
    global enable_debug_mode;
    auto enable_debug_mode = false;
    debug_function();
    auto enable_debug_mode = true;
    debug_function();
    std::cout << "\n--- Composed Decorators ---" << std::endl;
    global _attempt_count;
    auto _attempt_count = 0;
    auto result = complex_operation(2);
    std::cout << to_str("Complex operation result: ") << to_str(result) << std::endl;
}


int main() {
    /* docstring */
    std::cout << "=== Advanced Python Decorator Patterns ===" << std::endl;
    test_class_decorators();
    test_parameterized_decorators();
    test_advanced_patterns();
    std::cout << "\n=== All decorator tests completed successfully! ===" << std::endl;
    return 0;
}


